## Оглавление

1. [Описание работы](#start)
3. [Обзор данных](#preprocessing)
    * [Загрузка библиотек, подключение к базам данных](#3.1)
    * [Просмотр таблицы books](#3.2)
    * [Просмотр таблицы authors](#3.3)
    * [Просмотр таблицы publishers](#3.4)
    * [Просмотр таблицы ratings](#3.5)
    * [Просмотр таблицы reviews](#3.6)
3. [Исследование](#eda)
    * [Кол-во книг вышедших после 1 января 2000 года](#4.1)
    * [Кол-во обзоров и средняя оценка для каждой книги](#4.2)
    * [Издательства, выпустившие наибольшее кол-во книг более 50 страниц](#4.3)
    * [Автор, с самой высокой средней оценкой (книги с 50 и более оценками)](#4.4)
    * [Среднее кол-во обзоров от пользователей, которые поставили больше 50 оценок](#4.5)
6. [Вывод](#conclusion)

## <a id="start">Описание работы</a>

**Контекст**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задание**

 - Посчитайте, сколько книг вышло после 1 января 2000 года;
 - Для каждой книги посчитайте количество обзоров и среднюю оценку;
 - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
 - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
 - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Описание данных**

Таблица books
<br>Содержит данные о книгах:
 - book_id — идентификатор книги;
 - author_id — идентификатор автора;
 - title — название книги;
 - num_pages — количество страниц;
 - publication_date — дата публикации книги;
 - publisher_id — идентификатор издателя.

Таблица authors
<br>Содержит данные об авторах:
 - author_id — идентификатор автора;
 - author — имя автора.

Таблица publishers
<br>Содержит данные об издательствах:
 - publisher_id — идентификатор издательства;
 - publisher — название издательства.

Таблица ratings
<br>Содержит данные о пользовательских оценках книг:
 - rating_id — идентификатор оценки;
 - book_id — идентификатор книги;
 - username — имя пользователя, оставившего оценку;
 - rating — оценка книги.

Таблица reviews
<br>Содержит данные о пользовательских обзорах на книги:
 - review_id — идентификатор обзора;
 - book_id — идентификатор книги;
 - username — имя пользователя, написавшего обзор;
 - text — текст обзора.

## <a id="preprocessing">Обзор данных</a>

### <a id="3.1">Загрузка библиотек, подключение к базам данных</a>

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': '*******', # имя пользователя
						'pwd': '*************', # пароль
						'host': '***************',
						'port': **************, # порт подключения
						'db': '******************'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
											db_config['pwd'],
											db_config['host'],
											db_config['port'],
											db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
def call_result(call):
	return pd.io.sql.read_sql(call, con = engine)

### <a id="3.2">Просмотр таблицы books</a>

In [4]:
call = \
'''SELECT * 
FROM books
LIMIT 5'''
call_result(call)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### <a id="3.3">Просмотр таблицы authors</a>

In [5]:
call = \
'''SELECT * 
FROM authors
LIMIT 5'''
call_result(call)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### <a id="3.4">Просмотр таблицы publishers</a>

In [6]:
call = \
'''SELECT * 
FROM publishers
LIMIT 5'''
call_result(call)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### <a id="3.5">Просмотр таблицы ratings</a>

In [7]:
call = \
'''SELECT * 
FROM ratings
LIMIT 5'''
call_result(call)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### <a id="3.6">Просмотр таблицы reviews</a>

In [8]:
call = \
'''SELECT * 
FROM reviews
LIMIT 5'''
call_result(call)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## <a id="eda">Исследование</a>

### <a id="4.1">Кол-во книг вышедших после 1 января 2000 года</a>

In [9]:
call = \
'''SELECT COUNT (book_id) as books_count
FROM books
WHERE publication_date >= '2000-01-01'
'''
call_result(call)

books_count
0          821

### <a id="4.2">Кол-во обзоров и средняя оценка для каждой книги</a>

In [10]:
call = \
'''SELECT
  books.title,
  COUNT (distinct review_id) as review_count,
  ROUND(AVG(rating)::numeric, 2) AS avg_rating
FROM
  books
LEFT JOIN
  reviews ON reviews.book_id = books.book_id
LEFT JOIN
  ratings ON ratings.book_id = books.book_id
GROUP BY
  books.title
ORDER BY 
    COUNT(reviews.review_id) DESC'''
call_result(call)

title  review_count  \
0                              Twilight (Twilight  #1)             7   
1                  The Hobbit  or There and Back Again             6   
2                               The Catcher in the Rye             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...             6   
4    Harry Potter and the Chamber of Secrets (Harry...             6   
..                                                 ...           ...   
994  Disney's Beauty and the Beast (A Little Golden...             0   
995                          Essential Tales and Poems             0   
996                               Leonardo's Notebooks             0   
997                            The Natural Way to Draw             0   
998   The Cat in the Hat and Other Dr. Seuss Favorites             0   

     avg_rating  
0          3.66  
1          4.13  
2          3.83  
3          4.41  
4          4.29  
..          ...  
994        4.00  
995        4.00  
996        4.00  
997        3.00  
998        5.00  

[999 rows x 3 columns]

### <a id="4.3">Издательства, выпустившие наибольшее кол-во книг более 50 страниц</a>

In [11]:
call = \
'''SELECT
  publisher,
  COUNT (book_id) as books_count
FROM
  books
LEFT JOIN
  publishers ON publishers.publisher_id=books.publisher_id
WHERE
  num_pages>50
GROUP BY
  publisher
ORDER BY
  books_count DESC
LIMIT 1'''
call_result(call)

publisher  books_count
0  Penguin Books           42

### <a id="4.4">Автор, с самой высокой средней оценкой (книги с 50 и более оценками)</a>

In [12]:
call = \
'''SELECT
  author,
  ROUND(AVG(sub.avg_rating)::numeric, 2) AS high_avg_rating
FROM
(SELECT
  author,
  books.book_id,
  AVG(rating) AS avg_rating
FROM 
  authors
LEFT JOIN books ON books.author_id = authors.author_id
LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
  author,
  books.book_id
HAVING
  COUNT(rating) > 50) AS sub
GROUP BY
  author
ORDER BY
  high_avg_rating DESC
LIMIT 1'''
call_result(call)

author  high_avg_rating
0  J.K. Rowling/Mary GrandPré             4.28

### <a id="4.5">Среднее кол-во обзоров от пользователей, которые поставили больше 50 оценок</a>

In [13]:
call = \
'''SELECT
  ROUND(AVG(review_cnt)::numeric) as avg_review_cnt
FROM
(SELECT
  COUNT(review_id) as review_cnt
FROM
  reviews
WHERE username IN
(SELECT
  username
FROM
  ratings
GROUP BY
  username
HAVING
  COUNT(ratings) > 50)
GROUP BY
  username) as sub_review_cnt'''
call_result(call)

avg_review_cnt
0            24.0

## <a id="conclusion">Вывод</a>

 - После 1 января 2000 года вышла 821 книга;
 - Средняя оценка для каждой книги находится в диапазоне от 3 до 5 баллов;
 - Издательством выпустившим самое большое кол-во книг является Penguin Books, Vintage;
 - Автором с самой высокой средней оценкой 4.28 балла является J.K. Rowling / Mary GrandPré;
 - Пользователи, которые ставят более 50 оценок, в среднем пишут 24 обзора.